src/nvim/sign.c

```diff
@@ -77,736 +59,245 @@ static char *cmds[] = {
-                                      ? ""
-                                      : sign->se_group->sg_name));
-  tv_dict_add_nr(d,  S_LEN("lnum"), sign->se_lnum);
-  tv_dict_add_str(d, S_LEN("name"), sign_typenr2name(sign->se_typenr));
-  tv_dict_add_nr(d,  S_LEN("priority"), sign->se_priority);
-
-  return d;
-}
-
-// Sort the signs placed on the same line as "sign" by priority.  Invoked after
-// changing the priority of an already placed sign.  Assumes the signs in the
-// buffer are sorted by line number and priority.
-static void sign_sort_by_prio_on_line(buf_T *buf, sign_entry_T *sign)
-  FUNC_ATTR_NONNULL_ALL
-{
-  // If there is only one sign in the buffer or only one sign on the line or
-  // the sign is already sorted by priority, then return.
-  if ((sign->se_prev == NULL
-       || sign->se_prev->se_lnum != sign->se_lnum
-       || sign->se_prev->se_priority > sign->se_priority)
-      && (sign->se_next == NULL
-          || sign->se_next->se_lnum != sign->se_lnum
-          || sign->se_next->se_priority < sign->se_priority)) {
-    return;
-  }
-
-  // One or more signs on the same line as 'sign'
-  // Find a sign after which 'sign' should be inserted
-
-  // First search backward for a sign with higher priority on the same line
-  sign_entry_T *p = sign;
-  while (p->se_prev != NULL
-         && p->se_prev->se_lnum == sign->se_lnum
-         && p->se_prev->se_priority <= sign->se_priority) {
-    p = p->se_prev;
-  }
-  if (p == sign) {
-    // Sign not found. Search forward for a sign with priority just before
-    // 'sign'.
-    p = sign->se_next;
-    while (p->se_next != NULL
-           && p->se_next->se_lnum == sign->se_lnum
-           && p->se_next->se_priority > sign->se_priority) {
-      p = p->se_next;
-    }
-  }
-
-  // Remove 'sign' from the list
-  if (buf->b_signlist == sign) {
-    buf->b_signlist = sign->se_next;
-  }
-  if (sign->se_prev != NULL) {
-    sign->se_prev->se_next = sign->se_next;
-  }
-  if (sign->se_next != NULL) {
-    sign->se_next->se_prev = sign->se_prev;
-  }
-  sign->se_prev = NULL;
-  sign->se_next = NULL;
-
-  // Re-insert 'sign' at the right place
-  if (p->se_priority <= sign->se_priority) {
-    // 'sign' has a higher priority and should be inserted before 'p'
-    sign->se_prev = p->se_prev;
-    sign->se_next = p;
-    p->se_prev = sign;
-    if (sign->se_prev != NULL) {
-      sign->se_prev->se_next = sign;
-    }
-    if (buf->b_signlist == p) {
-      buf->b_signlist = sign;
-    }
-  } else {
-    // 'sign' has a lower priority and should be inserted after 'p'
-    sign->se_prev = p;
-    sign->se_next = p->se_next;
-    p->se_next = sign;
-    if (sign->se_next != NULL) {
-      sign->se_next->se_prev = sign;
-    }
-  }
-}
-
-/// Add the sign into the signlist. Find the right spot to do it though.
+/// @param sp  sign properties
+static void buf_addsign(buf_T *buf, int id, char *group, int prio, linenr_T lnum, sign_T *sp)
```

Comment:
The pointer case below looks a bit risky. I think it's better to cast at caller instead.
```suggestion
static void buf_addsign(buf_T *buf, uint32_t id, char *group, int prio, linenr_T lnum,
                        const sign_T *sp)
```